### ISSUE TITLE
Recent containers-0.5.9.1 exhibits observably conditional API

### ISSUE THREAD

**Issue by hvr** [haskell/containers#373]
I took a closer look at `containers` API after @treeowl pinged me about GHC 8.2 inclusion, and I noticed something problematic; one important invariant is not met which is required by the "inductive step" of the PVP, namely:

- *The observable exposed API shall be a function of its package version*
  (see also this [FAQ entry](http://pvp.haskell.org/faq/#what-implications-does-the-pvp-have-for-conditional-apis))

Specifically, I'm able to observe this conditionality quite easily by the following simple Haskell2010 program:

```hs
{-# LANGUAGE Haskell2010 #-}

module Main where

import qualified Data.Sequence as Seq

foo :: Seq.Seq ()
foo = Seq.Empty

main :: IO ()
main = print foo
```

Which compiles successfully with GHC 8.0, but fails with GHC 7.10:

```
Main.hs:8:7:
    Not in scope: data constructor ‘Seq.Empty’
    Perhaps you meant one of these:
      ‘Seq.EmptyL’ (imported from Data.Sequence),
      ‘Seq.EmptyR’ (imported from Data.Sequence),
      variable ‘Seq.empty’ (imported from Data.Sequence)
```

Even though it specified its direct dependencies according to the PVP contract, by declaring `build-depends: containers >= 0.5.9.1 && < 0.6`. This is particularly important given `containers` is a heavily depended core package in our ecosystem, and ought to be as compatible as possible to the PVP.

Long story short, for the existing `containers-0.5.9.1` release I suggest to retrofit a `base >= 4.9` constraint to make sure that  the API is no longer observably conditional in this respect.

**Comment by treeowl** [comment]
That's a big break, with no apparent path to resolution. Yes, that program
looks like Haskell 2010, but in fact it uses a pattern synonym. There is no
way *ever* to make that work with earlier GHC versions, let alone other
compilers. To satisfy that PVP requirement strictly, the API would have to
be split between containers and a hypothetical containers-patterns or some
such. I don't see how that could possibly be worth the trouble.

On Dec 22, 2016 3:15 AM, "Herbert Valerio Riedel" <notifications@github.com>
wrote:

I took a closer look at containers API after @treeowl
<https://github.com/treeowl> pinged me about GHC 8.2 inclusion, and I
noticed something problematic; one important invariant is not met which is
required by the "inductive step" of the PVP, namely:

   - *The observable exposed API shall be a function of its package version*
   (see also this FAQ entry
   <http://pvp.haskell.org/faq/#what-implications-does-the-pvp-have-for-conditional-apis>
   )

Specifically, I'm able to observe this conditionality quite easily by the
following simple Haskell2010 program:

{-# LANGUAGE Haskell2010 #-}
module Main where
import qualified Data.Sequence as Seq
foo :: Seq.Seq ()
foo = Seq.Empty
main :: IO ()
main = print foo

Which compiles successfully with GHC 8.0, but fails with GHC 7.10:

Main.hs:8:7:
    Not in scope: data constructor ‘Seq.Empty’
    Perhaps you meant one of these:
      ‘Seq.EmptyL’ (imported from Data.Sequence),
      ‘Seq.EmptyR’ (imported from Data.Sequence),
      variable ‘Seq.empty’ (imported from Data.Sequence)

Even though it specified its direct dependencies according to the PVP
contract, by declaring build-depends: containers >= 0.5.9.1 && < 0.6. This
is particularly important given containers is a heavily depended core
package in our ecosystem, and ought to be as compatible as possible to the
PVP.

Long story short, for the existing containers-0.5.9.1 release I suggest to
retrofit a base >= 4.9 constraint to make sure that the API is no longer
observably conditional in this respect.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/haskell/containers/issues/373>, or mute the thread

**Comment by treeowl** [comment]
I should point out that there are similar (or more egregious) violations
elsewhere. For example, some packages export particular functions only on
Unix or only on Windows.

**Comment by treeowl** [comment]
The only other approach would be to export the pattern synonyms using 7.10
syntax, forcing users to import them separately. The trouble is that we'd
have to stick with that for another bunch of years, annoying everyone.

**Comment by hvr** [comment]
@treeowl fair enough; lemme think of a scheme/workflow to make sure people don't rely on the conditional API parts w/o the necessary precautions; I'll get back to you

**Comment by treeowl** [comment]
@hvr, have you had further thoughts?

**Comment by jwaldmann** [comment]
It very unfortunate that container-0.5.9.1's `Empty` pattern breaks `lens-4.15.*`, so they lowered their upper bound to exclude the current containers package ( https://github.com/ekmett/lens/issues/698 )

Should I be worried? E.g., does containers 0.5.9 bring performance benefits that I would miss when using 0.5.8 (the last version without the pattern)?

**Comment by treeowl** [comment]
Hopefully lens will make a release compatible with containers 0.5.9.1
sometime soon. There are some performance improvements, but not as big in
most cases as some other recent releases.

**Comment by hvr** [comment]
@treeowl I forgot to get back to you about this. I did consult with a few parties and came to the conclusion that one can define a class of special cases (of which this is one instance) - and when it's not easy to avoid - where it makes sense to allow the API signature to become a function of not only the package version but also the available compiler features. I'm going to formalise this a bit better, but as far as `containers` is concerned, a clear signpost in the documentation totally suffices; the intent is so that API consumers get a chance to know how to write their dependency specifications by looking at `containers` documentation (we can't yet express this via the `@since` annotations so it has to be some free form annotation for now).

**Comment by sjakobi** [comment]
There hasn't been any activity here for 3 years. Can this be closed?!

**Comment by treeowl** [comment]
@sjakobi, you want to check if there's a line to document it? If so, shut this down.

**Comment by sjakobi** [comment]
> you want to check if there's a line to document it?

I don't see any documentation of this issue in the haddocks.

**Comment by meooow25** [comment]
This seems obsolete given that GHC <8.0 has been dropped.
